include/highfive/H5DataType.hpp

```diff
@@ -80,6 +87,68 @@ class AtomicType : public DataType {
 
     typedef T basic_type;
 };
+
+///
+/// \brief Create a compound HDF5 datatype
+///
+class CompoundType : public DataType {
```

Comment:
## CompoundType API changes proposal

### What about an immutable class?

`addMembers` and `addMember` member methods doesn't seem necessary to me. I would rather remove them and make the `CompoundType` class immutable. Something like:

```c++
class CompoundType : public DataType {
public:
  /// \brief Use for defining a sub-type of compound type
  struct member_def {
    std::string name;
    DataType base_type;
    size_t offset;
  };

  // allow instantiation of a CompoundType programmatically
  // like `addMember` and `addMembers`
  CompoundType(std::vector<member_def> t_members)
      : members(std::move(t_members))
      , size(t_size) {}

  CompoundType(std::initializer_list<member_def> t_members)
  : CompoundType({t_members}, t_size) {}

  void autoCreate();
  void manualCreate();
  void commit(const Object& object, const std::string& name);

  const std::vector<member_def> members;
};

CompoundType automatic_type{
    {"foo", AtomicType<int>{}},
    {"bar", CompoundType{{"pika", AtomicType<double>{}, 42}}}};
```

### What about one unique `create` method?

Regarding `manualCreate` and `autoCreate`, what about one unique `create` member method accepting an optional size in parameter?

### Actually ... what about removing the `create` method?

It is my understanding that the `create` method must always be called at least once. If this is indeed the case, I would:
1. pass the optional `size` in the class constructor
1. call the `create` member method in the constructor
1. mark the `create` member method private.

Something like this:
```c++
class CompoundType : public DataType {
public:
  ///
  /// \brief Use for defining a sub-type of compound type
  struct member_def {
    std::string name;
    DataType base_type;
    size_t offset;
  };

  // allow instantiation of a CompoundType programmatically
  // like `addMember` and `addMembers`
  CompoundType(std::vector<member_def> t_members)
      : members(std::move(t_members))
      , size(t_size) {}

  CompoundType(std::initializer_list<member_def> t_members)
  : CompoundType({t_members}, t_size) {}

  // 0 means size is calculated automatically
  void create(size_t size = 0);
  void commit(const Object& object, const std::string& name);

  const std::vector<member_def> members;
  const size_t size;
};

CompoundType automatic_type{
    {"foo", AtomicType<int>{}},
    {"bar", CompoundType{{"pika", AtomicType<double>{}, 42}}}};

CompoundType
    manual_type({{"foo", AtomicType<int>{}},
                 {"bar", CompoundType{{"pika", AtomicType<double>{}, 42}}}},
                46);
```

### Introspection

The class should provide a way to access the `members` member variable.
If the immutability proposal is accepted, then the `members` member variable can be `public` because `const`. Otherwise, there should be a `getMembers() const noexcept` member function.